packages/codemirror-blocks/src/ui/PrimitiveList.tsx

```diff
@@ -18,65 +18,64 @@ require("./PrimitiveList.less");
 type BasePrimitiveProps = {
   primitive: LanguagePrimitive;
   className: string;
-  onFocus: (primitive: LanguagePrimitive) => void;
-  onKeyDown: (e: React.KeyboardEvent) => void;
+  onFocus: (e: React.FocusEvent) => void;
 };
 
-export const Primitive = (props: BasePrimitiveProps) => {
-  const { primitive, className, onFocus, onKeyDown } = props;
+// @pcardune - I still needed to cast the ref here, in order to satisfy TS
+export const Primitive = React.forwardRef<HTMLElement, BasePrimitiveProps>(
+  (props, ref: React.RefObject<HTMLElement>) => {
+    const { primitive, className, onFocus } = props;
+    const focusedNode = useSelector(selectors.getFocusedNode);
+    const [_, connectDragSource, connectDragPreview] = useDrag({
+      type: ItemTypes.NODE,
+      item: { content: primitive.name },
+    });
 
-  const focusedNode = useSelector(selectors.getFocusedNode);
-
-  const [_, connectDragSource, connectDragPreview] = useDrag({
-    type: ItemTypes.NODE,
-    item: { content: primitive.name },
-  });
-
-  const handleKeyDown = (e: React.KeyboardEvent) => {
-    switch (defaultKeyMap[CodeMirror.keyName(e)]) {
-      case "Copy": {
-        e.preventDefault();
-        const node = primitive.getASTNode();
-        copy({ focusedNode }, [node]);
-        say("copied " + primitive.toString());
-        primitive.element?.focus(); // restore focus
-        return;
+    const handleKeyDown = (e: React.KeyboardEvent) => {
+      switch (defaultKeyMap[CodeMirror.keyName(e)]) {
+        case "Copy": {
+          e.preventDefault();
+          const node = primitive.getASTNode();
+          copy({ focusedNode }, [node]);
+          say("copied " + primitive.toString());
+          ref.current?.focus(); // restore focus
```

Comment:
Is this where you ran into the typescript problem? I don't think it's safe to use the ref that was passed down since it could be a ref callback function and not a ref object.

But why does focus need to be restored here? `say()` and `copy()` should not change focus. And if they do for implementation detail issues, then they should also restore focus themselves.

Reply:
This is an excellent idea. Just pushed a fix.